### ISSUE TITLE
Dragon cannot easily activate the file upload button

### ISSUE THREAD

**Issue by selfthinker** [alphagov/govuk-design-system#4031]
This issue is from the accessibility audit of GOV.UK Frontend components by DAC in July 2024.

## DAC's description

* Page: [File upload](https://design-system.service.gov.uk/components/file-upload/)
* WCAG reference: none (usabililty)
* Issue ID: DAC_Usability_Upload_File_02

The ‘choose file’ input does not respond to any verbal command for users of voice activation software. This means that the user must use verbal keyboard commands such as ‘Press tab’ etc, in order to access the component.

<img width="531" alt="Screenshot of a page using the file upload component" src="https://github.com/user-attachments/assets/4e72db3b-913f-456e-8118-1be2d4598a94">

### Voice activation user comments:

“I am unable to upload a photo with Dragon commands. The ‘choose file’ component does not respond to any Dragon commands. This means I must use keyboard commands with dragon to activate it this will simulate a manual keyboard through voice activation. I will use the command ‘press tab’ to move my focus onto the component. Then the command ‘press enters’ to activate for the file explorer to appear. This is very time consuming and frustrating to be able to access one component.
I would expect the ‘choose file’ button to respond to the command ‘click button’ and label by name. When I use the command ‘click choose file’ I would expect the file explore to automatically appear. This makes it more accessible for me to complete with dragon commands.”

## DAC's proposed solution

Consider using the following implementation to enable all users to use the file upload as
expected:

```html
<div>
  <span>
    <input type="file" id="file-upload" style="display: none;">
  </span>
  <label for="file-name">Upload your photo</label>
  <button id="file-upload-button" style="background:green; color:white;">Choose File</button>
  <input type="text" disabled readonly id="file-name" value="No file chosen">
</div>
```
```js
<script>
  document.getElementById("file-upload-button").addEventListener("click", function(){
    document.getElementById("file-upload").click();
    document.getElementById("file-name").value = document.getElementById("file-upload").value;
  });
</script>
```

## Other potential solutions

This is a known issue with Dragon. This happens with all native HTML `<input type="file">` elements, irrespective of implementation.

We need to discuss and explore this further first before deciding on a solution.
We might need to change the guidance to refer to the known issues.
Another option is to lobby for making Nuance fix the bug in Dragon.

## Related issues

* https://github.com/alphagov/reported-bugs/issues/35
* https://github.com/alphagov/govuk-frontend/issues/3686
* https://github.com/alphagov/govuk-design-system-backlog/issues/49
* https://github.com/alphagov/govuk-frontend/issues/2425
* https://github.com/alphagov/govuk-frontend/issues/3685

## Additional instances

This is the only instance of the file upload component.

## Needed roles

Frontend developer, content designer

**Comment by selfthinker** [comment]
We [investigated this](https://github.com/alphagov/govuk-frontend/issues/5309) and are confident we can mostly fix this.

We've [built a prototype](https://github.com/alphagov/govuk-frontend/pull/5305) to enhance the file upload component not just to fix this but also to add other useful enhancements. Because that was mostly successful, we plan to [build it properly](https://github.com/alphagov/govuk-design-system/issues/4242).

Although that generally fixes the issue, we noticed it only works the first time you interact with the component and after that it fails to react. We’ve looked at lots of different implementations across the web, and none of them fixes it consistently for Dragon.

The issue is that Chrome has a security feature which disallows any JavaScript that appears to do potentially malicious actions. Opening the file picker is one of the [features gated by user activation](https://developer.mozilla.org/en-US/docs/Web/Security/User_activation). When trying to do that from the second time onwards, there is an error message in the browser console which says:
“File chooser dialog can only be shown with a user activation”

We [asked DAC for their opinion](https://github.com/alphagov/govuk-design-system/issues/4194) on this, and they said that the majority of users wouldn't need to use the file upload more than once, so are not likely to notice that it won't work the second time.
And there is a work around to make it work the second time. When you say 'mouse click' (or any other command that interacts with the page) in between clicking on the file input button, it keeps on working.
I have added [some of the solutions with technical details to this JSbin](https://output.jsbin.com/zupanah), including testable code.

There is another potential solution to add a `role="button"` to the `<input type=file />`. That makes the "click button" command work but not "click choose file". This is very likely to cause more issues with screen readers, though, and is therefore not something we will use.
But we could also [create a browser extension to add it](https://github.com/alphagov/govuk-frontend/issues/5448) for Dragon users.

**Comment by selfthinker** [comment]
We've fixed this today by publishing an [improved version of the file upload component](https://design-system.service.gov.uk/components/file-upload/#using-the-improved-file-upload-component) in [version 5.9.0](https://github.com/alphagov/govuk-frontend/releases/tag/v5.9.0). Dragon users can now say any of the commands they would expect to work.

However, due to [browser security features](https://developer.mozilla.org/en-US/docs/Web/Security/User_activation), this only works the first time they interact with the component. If the component needs to be used more than once, users have to interact with the page in other ways first, like scrolling up or down or saying "mouse click" or similar things, or they need to revert to how they interacted with the broken component before.

The improvements to the component are only enabled if you use the `javascript` macro option or extra markup in your HTML. That means that services updating to the latest version will not automatically get those improvements.

PR that fixed this: https://github.com/alphagov/govuk-frontend/pull/5305
Epic for the work: https://github.com/alphagov/govuk-frontend/issues/5626
